src/providers/wfs/qgsbackgroundcachedfeatureiterator.cpp

```diff
@@ -0,0 +1,870 @@
+
+// -------------------------
+
+
+QgsFeatureDownloaderProgressDialog::QgsFeatureDownloaderProgressDialog( const QString &labelText,
+    const QString &cancelButtonText,
+    int minimum, int maximum, QWidget *parent )
+  : QProgressDialog( labelText, cancelButtonText, minimum, maximum, parent )
+{
+  mCancel = new QPushButton( cancelButtonText, this );
+  setCancelButton( mCancel );
+  mHide = new QPushButton( tr( "Hide" ), this );
+  connect( mHide, &QAbstractButton::clicked, this, &QgsFeatureDownloaderProgressDialog::hideRequest );
+}
+
+void QgsFeatureDownloaderProgressDialog::resizeEvent( QResizeEvent *ev )
+{
+  QProgressDialog::resizeEvent( ev );
+  // Note: this relies heavily on the details of the layout done in QProgressDialogPrivate::layout()
+  // Might be a bit fragile depending on QT versions.
+  QRect rect = geometry();
+  QRect cancelRect = mCancel->geometry();
+  QRect hideRect = mHide->geometry();
+  int mtb = style()->pixelMetric( QStyle::PM_DefaultTopLevelMargin );
+  int mlr = std::min( width() / 10, mtb );
+  if ( rect.width() - cancelRect.x() - cancelRect.width() > mlr )
+  {
+    // Force right alighnment of cancel button
+    cancelRect.setX( rect.width() - cancelRect.width() - mlr );
+    mCancel->setGeometry( cancelRect );
+  }
+  mHide->setGeometry( rect.width() - cancelRect.x() - cancelRect.width(),
+                      cancelRect.y(), hideRect.width(), cancelRect.height() );
+}
+
+// -------------------------
+
+QgsFeatureDownloaderImpl::QgsFeatureDownloaderImpl( QgsBackgroundCachedSharedData *shared, QgsFeatureDownloader *downloader ): mSharedBase( shared ), mDownloader( downloader )
+{
+  // Needed because used by a signal
+  qRegisterMetaType< QVector<QgsFeatureUniqueIdPair> >( "QVector<QgsFeatureUniqueIdPair>" );
+}
+
+QgsFeatureDownloaderImpl::~QgsFeatureDownloaderImpl()
+{
+  if ( mProgressDialog )
+    mProgressDialog->deleteLater();
+}
+
+void QgsFeatureDownloaderImpl::emitFeatureReceived( QVector<QgsFeatureUniqueIdPair> features )
+{
+  emit mDownloader->featureReceived( features );
+}
+
+void QgsFeatureDownloaderImpl::emitFeatureReceived( int featureCount )
+{
+  emit mDownloader->featureReceived( featureCount );
+}
+
+void QgsFeatureDownloaderImpl::emitEndOfDownload( bool success )
+{
+  emit mDownloader->endOfDownload( success );
+}
+
+void QgsFeatureDownloaderImpl::stop()
+{
+  QgsDebugMsgLevel( QStringLiteral( "QgsFeatureDownloaderImpl::stop()" ), 4 );
+  mStop = true;
+  emitDoStop();
+}
+
+void QgsFeatureDownloaderImpl::setStopFlag()
+{
+  QgsDebugMsgLevel( QStringLiteral( "QgsFeatureDownloaderImpl::setStopFlag()" ), 4 );
+  mStop = true;
+}
+
+void QgsFeatureDownloaderImpl::hideProgressDialog()
+{
+  mSharedBase->setHideProgressDialog( true );
+  mProgressDialog->deleteLater();
+  mProgressDialog = nullptr;
+}
+
+// Called from GUI thread
+void QgsFeatureDownloaderImpl::createProgressDialog( int numberMatched )
```

Comment:
I **really* don't like this gui specific logic sitting in here, it's just completely out of place. I would much prefer to see this replaced by a QgsTask.

Reply:
yeah... This is not new code. Just the existing one of the WFS provider I moved a bit between files. I don't really feel like going through the QgsTask road in this pull request though